prow/cmd/initupload/main.go

```diff
@@ -0,0 +1,152 @@
+	dryRun             = flag.Bool("dry-run", true, "do not interact with GCS")
+)
+
+func main() {
+	flag.Parse()
+
+	if !*dryRun {
+		if *gcsBucket != "" {
+			logrus.Fatal("No GCS bucket specified")
+		}
+
+		if *gceCredentialsFile != "" {
+			logrus.Fatal("No GCE credentials specified")
+		}
+	}
+
+	if *pathStrategy != "legacy" && *pathStrategy != "explicit" && *pathStrategy != "single" {
+		logrus.Fatal("Path strategy must be one of 'legacy', 'explicit', or 'single'")
+	}
+
+	if *pathStrategy != "explicit" && (*defaultOrg == "" || *defaultRepo == "") {
+		logrus.Fatalf("Default org and repo must be provided for strategy %q", *pathStrategy)
+	}
+
+	var builder gcs.RepoPathBuilder
+	switch *pathStrategy {
+	case "explicit":
+		builder = gcs.NewExplicitRepoPathBuilder()
+	case "legacy":
+		builder = gcs.NewLegacyRepoPathBuilder(*defaultOrg, *defaultRepo)
+	case "single":
+		builder = gcs.NewSingleDefaultRepoPathBuilder(*defaultOrg, *defaultRepo)
+	}
+
+	spec, err := pjutil.ResolveSpecFromEnv()
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not resolve job spec")
+	}
+
+	gcsPath := gcs.PathForSpec(spec, builder)
+	if *subDir != "" {
+		gcsPath = path.Join(gcsPath, *subDir)
+	}
+
+	var cloneRecords []clone.Record
+	data, err := ioutil.ReadFile(*cloneLog)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not read clone log")
+	}
+	err = json.Unmarshal(data, &cloneRecords)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not unmarshal clone records")
+	}
+
+	failed := false
+	buildLog := bytes.Buffer{}
+	for _, record := range cloneRecords {
+		buildLog.WriteString(clone.FormatRecord(record))
+		failed = failed || record.Failed
+	}
+
+	uploadTargets := map[string]gcs.UploadFunc{
+		path.Join(gcsPath, "build-log.txt"): gcs.DataUpload(&buildLog),
+	}
+
+	started := struct {
+		Timestamp int64 `json:"timestamp"`
+	}{
+		Timestamp: time.Now().Unix(),
+	}
+	startedData, err := json.Marshal(&started)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not marshal starting data")
+	} else {
+		uploadTargets[path.Join(gcsPath, "started.json")] = gcs.DataUpload(bytes.NewBuffer(startedData))
+	}
+
+	if failed {
+		finished := struct {
+			Timestamp int64 `json:"timestamp"`
+			Passed    bool  `json:"passed"`
+		}{
+			Timestamp: time.Now().Unix(),
+			Passed:    false,
+		}
+		finishedData, err := json.Marshal(&finished)
+		if err != nil {
+			logrus.WithError(err).Fatal("Could not marshal finishing data")
+		} else {
+			uploadTargets[path.Join(gcsPath, "finished.json")] = gcs.DataUpload(bytes.NewBuffer(finishedData))
+		}
+	}
+
+	if !*dryRun {
+		ctx := context.Background()
+		gcsClient, err := storage.NewClient(ctx, option.WithCredentialsFile(*gceCredentialsFile))
+		if err != nil {
+			logrus.WithError(err).Fatal("Could not connect to GCS")
+		}
+
+		if err := gcs.UploadToGcs(gcsClient.Bucket(*gcsBucket), uploadTargets); err != nil {
```

Comment:
`gcs.UploadToGcs()` seems a little redundant. What do you think of `gcs.Upload()`?

(note: `storage.NewClient()` above)